### PR TITLE
Support for registering micrometer registries with corfu + Dropwizard…

### PR DESCRIPTION
**Pull request by PavelZaytsev**
## Overview

- Custom registry integration for corfu metrics.
- Dropwizard registry is configured with sliding window reporting.

### CODE REVIEW

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/RegistryProvider.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.corfudb.common.metrics.micrometer.registries;
+
+import io.micrometer.core.instrument.MeterRegistry;
+
+/**
+ * Clients who are willing to integrate their registries into Corfu metrics composite registry
+ * must implement this interface.
+ */
```
Comment (nvaishampayan517) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/registries/RegistryProvider.java:8]: should MeterRegistryProvider implement this interface?

Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/registries/RegistryProvider.java:8]: No, MeterRegistryProvider contains the CompositeRegistry. The CompositeRegistry holds the implementations of different registries (LoggingRegistry, DropwizardMeterRegistry, PrometheusMeterRegistry, etc). That way it pushes the same metrics to different backends simultaneously (this is what we want). RegistryProvider is the service provider interface, it is meant to be extended by the third parties. The parties implement this interface and provide Corfu their own configured registries. Corfu adds these registries to a CompositeRegistry and pushes all the metrics to it.

File: common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java [hunk 2]
```diff
@@ -51,24 +55,84 @@ public static synchronized void init(Logger logger, Duration loggingInterval, St
...
      * Get the previously configured meter registry.
      * If the registry has not been previously configured, return an empty option.
      *
      * @return An optional configured meter registry.
      */
     public static synchronized Optional<MeterRegistry> getInstance() {
-        return meterRegistry;
+        return Optional.of(meterRegistry);
+    }
+
```
[the remark this answers is not in the document]
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java:128]: Fixed the description

File: common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java [hunk 3]
```diff
@@ -79,15 +142,13 @@ private static void init(Supplier<Optional<MeterRegistry>> meterRegistrySupplier
...
         if (!id.isPresent()) {
-            throw new IllegalStateException("Id must be present to deregister meters.");
+            log.warn("Id must be present to deregister meters.");
+            return;
         }
         String server = id.get();
         Tags tagsToLookFor = tags.and(Tag.of("id", server));
         Meter.Id id = new Meter.Id(name, tagsToLookFor, null, null, type);
-        meterRegistry.ifPresent(registry -> registry.remove(id));
+        meterRegistry.remove(id);
```
Comment (zhangn49) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java:152]: Local variable `id` has the same name with the global id?

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import io.micrometer.core.instrument.AbstractDistributionSummary;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.distribution.DistributionStatisticConfig;
+import io.micrometer.core.instrument.distribution.TimeWindowMax;
+import io.micrometer.core.instrument.util.MeterEquivalence;
+import io.micrometer.core.lang.Nullable;
+
+import java.util.concurrent.atomic.DoubleAdder;
+
+public class DropwizardDistributionSummaryCustom extends AbstractDistributionSummary {
```
Comment (zhangn49) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java:12]: Would be great to have a comment for this class

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardMeterRegistryWithSlidingTimeWindow.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import io.micrometer.core.instrument.util.HierarchicalNameMapper;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class overrides the timer and histogram of DropwizardMeterRegistry and adds a
+ * SlidingWindowReservoirArray. This allows for the reporting of measurements made within the last
+ * n TimeUnits. This is the default behavior in other micrometer registries.
+ */
+public class DropwizardMeterRegistryWithSlidingTimeWindow extends DropwizardMeterRegistry {
```
Comment (zhangn49) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardMeterRegistryWithSlidingTimeWindow.java:22]: Seems to be unused. Will it be used in the next PR?

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    private final com.codahale.metrics.Histogram impl;
```
Comment (zhangn49) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java:13]: Do we want to import the drop wizard package? DropwizardTimerCustom does import but here it doesn't. Would be great to be consistent.

**Review by zhangn49 — COMMENTED** [review]
minor comments

**Review by xnull — APPROVED** [review]
looks good

File: common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java (hunk 3 above)
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java:152]: Fixed

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java (hunk 4 above)
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java:12]: Renamed and added comments

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardMeterRegistryWithSlidingTimeWindow.java (hunk 5 above)
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardMeterRegistryWithSlidingTimeWindow.java:22]: It's used by the clients on their side within the RegistryProvider implementation.

File: common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java (hunk 6 above)
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/registries/dropwizard/DropwizardDistributionSummaryCustom.java:13]: fixed

**Review by zhangn49 — APPROVED** [review]
LGTM, please fix Codacy issues

**Pull request merged**